### PR TITLE
PI Modification: Terminate call activity

### LINKED ISSUE #9702: Modify a process instance by terminating an element instance of a call activity
## Description

We want to support the termination of an element instance that belongs to a call activity. As a result, the child process instance of the call activity, and the call activity itself should be terminated. 

All active element instances of the child process instance should be terminated as well (recursively). 

We should produce the same sequence of events as a regular termination (i.e. `call activity: terminating`, `child process instance: terminating`, ..., `child process instance: terminated`, `call activity: terminated`). 

For example:
image

* terminate the call activity (according to #9690)
* terminate the child process instance of the call activity 
* propagate the termination of the call activity to its flow scope (according to #9691)  

Blocked by 
- #9691

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

When a terminate instruction is targetting a call activity we need to make sure this call activity is terminated correctly. Not only must we terminate the element instance of the call activity, we also have to terminate the child process.

## Related issues

closes #9702 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: test-util/src/main/java/io/camunda/zeebe/test/util/record/ProcessInstanceRecordStream.java [hunk 1]
```diff
@@ -42,6 +42,14 @@ public ProcessInstanceRecordStream withProcessInstanceKey(final long processInst
...
   }
 
+  public ProcessInstanceRecordStream withProcessInstanceKeyOrParentProcessInstanceKey(
+      final long processInstanceKey) {
+    return valueFilter(
+        v ->
+            v.getProcessInstanceKey() == processInstanceKey
+                || v.getParentProcessInstanceKey() == processInstanceKey);
+  }
+
```
Comment (korthout) [inline test-util/src/main/java/io/camunda/zeebe/test/util/record/ProcessInstanceRecordStream.java:52]: ❓ Do we have any other places where we use a similar check that could be refactored to use this new method?

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java [hunk 2]
```diff
@@ -533,10 +533,15 @@ private void terminateElement(
...
     if (elementType == BpmnElementType.EVENT_SUB_PROCESS
-        || elementType == BpmnElementType.SUB_PROCESS) {
+        || elementType == BpmnElementType.SUB_PROCESS
+        || elementType == BpmnElementType.PROCESS) {
       elementInstanceState.getChildren(elementInstanceKey).stream()
           .filter(ElementInstance::canTerminate)
           .forEach(childInstance -> terminateElement(childInstance, sideEffects));
+    } else if (elementType == BpmnElementType.CALL_ACTIVITY) {
+      final var calledActivityElementInstanceState =
+          elementInstanceState.getInstance(elementInstance.getCalledChildInstanceKey());
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java:543]: 🔧 I think the variable name accidentally got postfixed with `State`, but it should probably be named `childProcessInstance` anyways

**Review by korthout — APPROVED** [review]
Nice work @remcowesterhoud 👍 LGTM

🔧 Please have a look at my suggestions before merging

File: test-util/src/main/java/io/camunda/zeebe/test/util/record/ProcessInstanceRecordStream.java (hunk 1 above)
Comment (remcowesterhoud) [reply test-util/src/main/java/io/camunda/zeebe/test/util/record/ProcessInstanceRecordStream.java:52]: Honestly, I didn't look. I can imagine it's a useful filter for other test cases, hence why I put it in here. It's less straightforward to use my IDE to find tests that could use it currently.

I tried some text searching and couldn't find anything, but obviously this method fails as soon as something is written different.

**Comment by remcowesterhoud** [comment]
bors merge

**Pull request merged**